### PR TITLE
Adds area editing, linking, and filling to buildmode

### PR DESCRIPTION
**Pull request by monster860**
Right click on shit to select it's area, left click to paint area. Right click on buildmode button to create a new area of whatever type you want.

Fill mode works the same as map gen, but you rightclick on buildmode button to select an object type

:cl: monster860
rscadd: Adds area editing, link mode, and fill mode to the buildmode tool
/:cl:

### CODE REVIEW

**Comment by Dumbdumn5** [comment]
I. Love. You.

File: code/modules/admin/buildmode.dm [hunk 1]
```diff
@@ -234,6 +250,20 @@
...
+				var/areaname = input(user,"Enter area name:", "Area name", "Area")
+				if(!areaname || !length(areaname))
+					return
+				storedarea = new areatype
+				storedarea.power_equip = 0
+				storedarea.power_light = 0
+				storedarea.power_environ = 0
+				storedarea.always_unpowered = 0
+				storedarea.name = areaname
+				areaimage.loc = storedarea
```
Comment (Crazylemon64) [inline code/modules/admin/buildmode.dm]: The reason for setting all these things to 0 is so that you can use VV to do that by hand, right?

Comment (Tastyfish) [inline code/modules/admin/buildmode.dm]: power_X is updated by APC processing, it was probably 0'd out here in case the APC is removed from the area.

**Comment by monster860** [comment]
Fills.

**Comment by Crazylemon64** [comment]
A fill mode! :D

File: code/modules/admin/buildmode.dm [hunk 2]
```diff
@@ -264,6 +270,19 @@
...
+			if(!ispath(objholder))
+				objholder = pick_closest_path(target_path)
+				if(!objholder || ispath(objholder, /area))
+					objholder = /obj/structure/closet
+					return
+			else
+				if(ispath(objholder,/mob) && !check_rights(R_DEBUG,0))
+					objholder = /obj/structure/closet
+			cornerA = null
+			cornerB = null
```
Comment (Crazylemon64) [inline code/modules/admin/buildmode.dm]: Might wanna use Reset() here

**Comment by monster860** [comment]
Yep. Everything crazylemon was going to do but didn't.

**Comment by Crazylemon64** [comment]
Only true if you specify that as everything I publicly SAID I was going to do, I also want to add a button editor for linking doors, too.

Also, that statement seems a bit confrontational, although I assume you mean the best, for sake of avoiding conflict.

File: code/modules/admin/buildmode.dm [hunk 3]
```diff
@@ -234,6 +256,33 @@
...
```
Comment (Crazylemon64) [inline code/modules/admin/buildmode.dm]: You may want to use `pick_closest_path` so they don't need to give the exact path to make something.

**Comment by monster860** [comment]
Added the changelog shit

**Comment by Fox-McCloud** [comment]
Going to make the suggestion @Crazylemon64 made?

**Comment by monster860** [comment]
Yeah, and if I'm doing it, might as well make it link everything.

**Comment by Crazylemon64** [comment]
Well, I wasn't explicitly suggesting that thing I was going to do, but I guess if you're doing that, that's rather helpful! I mean, I ALSO wanted to split buildmodes into separate datums, so we don't just have a bunch of switch statements for all the procs, but...

**Comment by monster860** [comment]
Link mode!

File: code/modules/admin/buildmode.dm [hunk 4]
```diff
@@ -90,6 +91,32 @@
...
+	M.Scale(1, sqrt((x_offset * x_offset) + (y_offset * y_offset)) / 32)
+	M.Turn(90 - Atan2(x_offset, y_offset)) // So... You pass coords in order x,y to this version of atan2. It should be called acsc2.
+	M.Translate(atom_a.pixel_x, atom_a.pixel_y)
+	
+	transform = M
+	cl = c
+	cl.images += I
+
+/obj/effect/buildmode_line/Destroy()
+	cl.images -= I
```
Comment (tigercat2000) [inline code/modules/admin/buildmode.dm]: you should wrap this
```
/obj/effect/buildmode_line/Destroy()
    if(I)
        if(istype(cl))
            cl.images -= I
            cl = null
        qdel(I)
        I = null
    return ..()
```

**Comment by Crazylemon64** [comment]
Hmm, I think I might want to do a shuttle-based buildmode, too...

File: code/modules/admin/buildmode.dm [hunk 5]
```diff
@@ -188,6 +229,22 @@
 			user << "<span class='notice'>Left Mouse Button on obj/turf/mob   = Spawn a Copy of selected target</span>"
 			user << "<span class='notice'>Right Mouse Button on obj/mob = Select target to copy</span>"
 			user << "<span class='notice'>***********************************************************</span>"
+		if(AREAEDIT_BUILDMODE)
+			user << "<span class='notice'>***********************************************************</span>"
```
Comment (tigercat2000) [inline code/modules/admin/buildmode.dm]: to_chat(user, "<span class='notice'>***********************************************************</span>")

File: code/modules/admin/buildmode.dm [hunk 6]
```diff
@@ -188,6 +229,22 @@
...
+			user << "<span class='notice'>Left Mouse Button on obj/turf/mob  = Paint area</span>"
```
Comment (tigercat2000) [inline code/modules/admin/buildmode.dm]: above, but for all of these lines

File: code/modules/admin/buildmode.dm [hunk 7]
```diff
@@ -396,3 +487,109 @@
...
+		if(FILL_BUILDMODE)
+			if(!cornerA)
+				cornerA = select_tile(get_turf(object))
+				return
+			if(cornerA && !cornerB)
+				cornerB = select_tile(get_turf(object))
+			if(left_click) //rectangular
+				if(cornerA && cornerB)
+					if(!objholder)
+						user << "<span class='warning'>Select object type first.</span>"
```
Comment (tigercat2000) [inline code/modules/admin/buildmode.dm]: to_chat

File: code/modules/admin/buildmode.dm [hunk 8]
```diff
@@ -396,3 +487,109 @@
...
+				if(istype(link_obj, /obj/machinery/door_control) && istype(object, /obj/machinery/door/airlock))
+					var/obj/machinery/door_control/M = link_obj
+					var/obj/machinery/door/airlock/P = object
+					if(!M.id || M.id == "")
+						M.id = input(holder, "Please select an ID for the button", "Buildmode", "")
+						if(!M.id || M.id == "")
+							goto(line_jump)
+					if(P.id_tag == M.id && (P in range(M.range, M)) && P.id_tag && P.id_tag != "")
+						P.id_tag = null
+						holder << "[P] unlinked."
```
Comment (tigercat2000) [inline code/modules/admin/buildmode.dm]: to_chat

File: code/modules/admin/buildmode.dm [hunk 9]
```diff
@@ -396,3 +487,109 @@
...
+				if(istype(link_obj, /obj/machinery/door_control) && istype(object, /obj/machinery/door/poddoor))
+					var/obj/machinery/door_control/M = link_obj
+					var/obj/machinery/door/poddoor/P = object
+					if(!M.id || M.id == "")
+						M.id = input(holder, "Please select an ID for the button", "Buildmode", "")
+						if(!M.id || M.id == "")
+							goto(line_jump)
+					if(P.id_tag == M.id && P.id_tag && P.id_tag != "")
+						P.id_tag = null
+						holder << "[P] unlinked."
```
Comment (tigercat2000) [inline code/modules/admin/buildmode.dm]: to_chat

File: code/modules/admin/buildmode.dm (hunk 2 above)
Comment (Fox-McCloud) [inline code/modules/admin/buildmode.dm]: You still need to do this, @monster860

File: code/modules/admin/buildmode.dm (hunk 3 above)
Comment (Fox-McCloud) [inline code/modules/admin/buildmode.dm]: You still need to do this, @monster860

Comment (monster860) [inline code/modules/admin/buildmode.dm]: Doesn't it already use `pick_closest_path` here though?

Comment (Crazylemon64) [inline code/modules/admin/buildmode.dm]: Oh! I see it! My bad! Now to make fox look like a crazy person!

**Comment by monster860** [comment]
There, @Fox-McCloud @TheDZD @tigercat2000 Lien nutes are fixed.

**Pull request merged**